### PR TITLE
[3.10] Clear buffered output at all levels

### PR DESCRIPTION
**Pull request by zero-24**
Pull Request for Issue #18088 

### Summary of Changes

Add back the check to clear the buffered output. cc @SharkyKZ 

### Testing Instructions

Enter non-existing url in browser to trigger 404 page.

### Actual result BEFORE applying this Pull Request

Default template error page should be displayed, as usual.

### Expected result AFTER applying this Pull Request

Default template error page should be displayed, as usual.

### Documentation Changes Required

None

### CODE REVIEW

**Comment by ethernidee** [comment]
The following code:
```php
// Clear buffered output at all levels
$callerFunction = static::getCallerFunctionName();

if ($callerFunction === false)
{
  while (ob_get_level())
  {
    ob_end_clean();
  }
}
```

should be replaced with:

```php
// Clear buffered output at all levels
while (ob_get_level())
{
  ob_end_clean();
}
```

**Comment by zero-24** [comment]
Ok will do when i'm back at my PC.

**Comment by zero-24** [comment]
Thanks @ethernidee this has just been commited to this PR.

**Comment by particthistle** [comment]
I have tested this item :white_check_mark: successfully on 0fda10a9e11ee0f04da2fbfb5a8bcafdfd087639<br /><br />Default template error page should be displayed, as expected.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30786">issues.joomla.org/tracker/joomla-cms/30786</a>.</sub>

**Comment by SharkyKZ** [comment]
Tests are failing.

**Comment by ethernidee** [comment]
1) JErrorPageTest::testEnsureTheErrorPageIsCorrectlyRendered
Failed asserting that false matches expected '<title>500 - Testing JErrorPage::render() with RuntimeException</title>Testing JErrorPage::render() with RuntimeException'.

/drone/src/tests/unit/suites/libraries/cms/error/JErrorPageTest.php:79

2) JErrorPageTest::testEnsureTheErrorPageIsCorrectlyRenderedWithThrowables
Failed asserting that false matches expected '<title>500 - Testing JErrorPage::render() with PHP 7 Error</title>Testing JErrorPage::render() with PHP 7 Error'.

/drone/src/tests/unit/suites/libraries/cms/error/JErrorPageTest.php:123

---------------------------------

Such things cannot be tested if buffer is cleared properly. The tests are invalid, imho.

**Comment by zero-24** [comment]
> Such things cannot be tested if buffer is cleared properly. The tests are invalid, imho.

What is the point of clearing the buffer there anyway?

**Comment by ethernidee** [comment]
Imagine you render Joomla module. First level of buffer capturing is on. Then in module template you write ob_start() for internal subwidget. If any error happens, Joomla will clear second level buffer and leave top of the default template rendered + will add error page html. This is totally wrong. All buffer levels should be cleared before rewriting the whole response, whether it's json or html response.

I faced such situations on working sites and was really surprised.

**Comment by zero-24** [comment]
Makes sense, thanks

**Comment by zero-24** [comment]
The latest commits make sure the test suite passes as it starts the output buffer right after we cleaded it again. Else that thing will fail as the output puffes is clreandend but not restarted.

**Comment by ethernidee** [comment]
Excellent solution!

**Comment by richard67** [comment]
Thanks!

File: libraries/src/Exception/ExceptionHandler.php [hunk 1]
```diff
@@ -176,7 +182,7 @@ public static function render($error)
 	protected static function getCallerFunctionName()
```
Comment (wilsonge) [inline libraries/src/Exception/ExceptionHandler.php:182]: @zero-24 I think this function can be removed now - it's not longer being called

Comment (zero-24) [reply libraries/src/Exception/ExceptionHandler.php:182]: it could be used as its part of the API right? It can be removed with 4 i guess right?

Comment (wilsonge) [reply libraries/src/Exception/ExceptionHandler.php:182]: It's a protected function and introduced in 3.10 - its only been in alpha's so far?

Comment (zero-24) [reply libraries/src/Exception/ExceptionHandler.php:182]: Ah ok will do a PR to get rid of it.

Comment (zero-24) [reply libraries/src/Exception/ExceptionHandler.php:182]: PR done: https://github.com/joomla/joomla-cms/pull/32175

**Pull request merged**